scipy/sparse/construct.py

```diff
@@ -664,16 +664,25 @@ def block_diag(mats, format=None, dtype=None):
            [0, 0, 0, 7]])
 
     """
-    nmat = len(mats)
-    rows = []
+    row = []
+    col = []
+    data = []
+    r_idx = 0
+    c_idx = 0
     for ia, a in enumerate(mats):
-        row = [None]*nmat
         if issparse(a):
-            row[ia] = a
+            a = a.tocsr()
         else:
-            row[ia] = coo_matrix(a)
-        rows.append(row)
-    return bmat(rows, format=format, dtype=dtype)
+            a = csr_matrix(a)
```

Comment:
Recasting dense arrays to csr here is completely unnecessay and a performance bottleneck for building a matrix from many small dense arrays. I was testing building a block-diagonal out of 20000 3x3 dense arrays, and 98% of the time is spent on this line, removing it results in huge speedup and there is no downside as far as I see.
```suggestion
            pass
```